File: static/include/examples/fleet-api/fleet-management-api-machine.go
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"go.viam.com/rdk/app"
	"go.viam.com/rdk/logging"
)

// Configuration constants – replace with your actual values
var (
	API_KEY     = "" // API key, find or create in your organization settings
	API_KEY_ID  = "" // API key ID, find or create in your organization settings
	ORG_ID      = "" // Organization ID, find or create in your organization settings
	EMAIL_ADDRESS = "" // Email address of the user to get the user id for
	LOCATION_ID = "" // Location ID, find or create in your organization settings
)

func main() {
	// :remove-start:
	ORG_ID = os.Getenv("TEST_ORG_ID")
	API_KEY = os.Getenv("VIAM_API_KEY")
	API_KEY_ID = os.Getenv("VIAM_API_KEY_ID")
	LOCATION_ID = "pg5q3j3h95"
	// :remove-end:
	logger := logging.NewDebugLogger("client")
	ctx := context.Background()

	// Make a ViamClient
	viamClient, err := app.CreateViamClientWithAPIKey(
		ctx, app.Options{}, API_KEY, API_KEY_ID, logger)
	if err != nil {
		logger.Fatal(err)
	}
	defer viamClient.Close()

	// Instantiate an AppClient called "cloud"
	// to run fleet management API methods on
	cloud := viamClient.AppClient()

	// Test constants
	MACHINE_ID := "5ec7266e-f762-4ea8-9c29-4dd592718b48"
	PART_ID := "deb8782c-7b48-4d35-812d-2caa94b61f77"

	// Get robot
	robot, err := cloud.GetRobot(ctx, MACHINE_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if robot.ID != MACHINE_ID {
		logger.Fatal("Robot ID mismatch")
	}

	// Get robot API keys
	apiKeys, err := cloud.GetRobotAPIKeys(ctx, MACHINE_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(apiKeys) < 1 {
		logger.Fatal("Expected at least 1 API key")
	}
	if apiKeys[0].Authorizations[0].ResourceID != MACHINE_ID {
		logger.Fatal("API key resource ID mismatch")
	}

	// Get robot parts
	listOfParts, err := cloud.GetRobotParts(ctx, MACHINE_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfParts) != 1 {
		logger.Fatal("Expected 1 robot part")
	}
	if listOfParts[0].ID != PART_ID {
		logger.Fatal("Robot part ID mismatch")
	}

	// Get robot part
	myRobotPart, _, err := cloud.GetRobotPart(ctx, PART_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if myRobotPart.ID != PART_ID {
		logger.Fatal("Robot part ID mismatch")
	}


    filter := ""
    pageToken := ""
    startTime := time.Now().Add(-720 * time.Hour)
    endTime := time.Now()
    limit := 5
    source := ""
	partLogs, _, err := cloud.GetRobotPartLogs(
        ctx,
        PART_ID,
        &app.GetRobotPartLogsOptions{
			Filter: &filter,
			PageToken: &pageToken,
			Levels: []string{"INFO", "WARN", "ERROR"},
			Start: &startTime,
			End: &endTime,
			Limit: &limit,
			Source: &source,
		})
	if err != nil {
		logger.Fatal(err)
	}
	if len(partLogs) != 5 {
		logger.Fatal("Expected 5 log entries")
	}

	// Tail robot part logs
	logFilter := "error"
	logsStream, err := cloud.TailRobotPartLogs(ctx, PART_ID, false, &app.TailRobotPartLogsOptions{
		Filter: &logFilter,
	})
	if err != nil {
		logger.Fatal(err)
	}
	if logsStream == nil {
		logger.Fatal("Logs stream should not be nil")
	}

	// Get robot part history
	partHistory, err := cloud.GetRobotPartHistory(ctx, PART_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(partHistory) < 1 {
		logger.Fatal("Expected at least 1 history entry")
	}

	// Create new robot part
	newPartID, err := cloud.NewRobotPart(ctx, MACHINE_ID, "new-part")
	if err != nil {
		logger.Fatal(err)
	}
	if newPartID == "" {
		logger.Fatal("New part ID should not be empty")
	}

	listOfParts, err = cloud.GetRobotParts(ctx, MACHINE_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfParts) != 2 {
		logger.Fatal("Expected 2 robot parts after creation")
	}

	robotConfig := map[string]interface{}{
		"components": []map[string]interface{}{
			{
				"name":       "camera-1",
				"api":        "rdk:component:camera",
				"model":      "rdk:builtin:fake",
				"attributes": map[string]interface{}{},
			},
		},
	}

	updatedPart, err := cloud.UpdateRobotPart(ctx, PART_ID, "test-part", robotConfig)
	if err != nil {
		logger.Fatal(err)
	}
	fmt.Println(updatedPart)
	if updatedPart.Name != "test-part" {
		logger.Fatal("Updated part name mismatch")
	}

	// Mark part as main
	err = cloud.MarkPartAsMain(ctx, newPartID)
	if err != nil {
		logger.Fatal(err)
	}

	newMachinePart, _, err := cloud.GetRobotPart(ctx, newPartID)
	if err != nil {
		logger.Fatal(err)
	}
	if !newMachinePart.MainPart {
		logger.Fatal("New machine part should be main part")
	}

	err = cloud.MarkPartAsMain(ctx, PART_ID)
	if err != nil {
		logger.Fatal(err)
	}

	oldMachinePart, _, err := cloud.GetRobotPart(ctx, PART_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if !oldMachinePart.MainPart {
		logger.Fatal("Old machine part should be main part")
	}

	// Mark part for restart
	err = cloud.MarkPartForRestart(ctx, newPartID)
	if err != nil {
		logger.Fatal(err)
	}

	// Create robot part secret
	partWithNewSecret, err := cloud.CreateRobotPartSecret(ctx, newPartID)
	if err != nil {
		logger.Fatal(err)
	}
	if partWithNewSecret == nil {
		logger.Fatal("Part with new secret should not be nil")
	}
	if partWithNewSecret.ID != newPartID {
		logger.Fatal("Part secret ID mismatch")
	}

	// Delete robot part secret
    fmt.Printf("Deleting robot part secret: %+v\n", partWithNewSecret.Secrets[0].ID)
	err = cloud.DeleteRobotPartSecret(ctx, newPartID, partWithNewSecret.Secrets[0].ID)
	if err != nil {
		logger.Fatal(err)
	}

	// Delete robot part
	err = cloud.DeleteRobotPart(ctx, newPartID)
	if err != nil {
		logger.Fatal(err)
	}

	listOfParts, err = cloud.GetRobotParts(ctx, MACHINE_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfParts) != 1 {
		logger.Fatal("Expected 1 robot part after deletion")
	}

	// List robots
	listOfMachines, err := cloud.ListRobots(ctx, LOCATION_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfMachines) != 1 {
		logger.Fatal("Expected 1 machine")
	}
	if listOfMachines[0].ID != MACHINE_ID {
		logger.Fatal("Machine ID mismatch")
	}

	// Create new robot
	newMachineID, err := cloud.NewRobot(ctx, "test-robot", LOCATION_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if newMachineID == "" {
		logger.Fatal("New machine ID should not be empty")
	}

	listOfMachines, err = cloud.ListRobots(ctx, LOCATION_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfMachines) != 2 {
		logger.Fatal("Expected 2 machines after creation")
	}

	// Update robot
	updatedMachine, err := cloud.UpdateRobot(ctx, newMachineID, "test-robot-new-name", LOCATION_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if updatedMachine.Name != "test-robot-new-name" {
		logger.Fatal("Updated machine name mismatch")
	}

	// Delete robot
	err = cloud.DeleteRobot(ctx, newMachineID)
	if err != nil {
		logger.Fatal(err)
	}

	listOfMachines, err = cloud.ListRobots(ctx, LOCATION_ID)
	if err != nil {
		logger.Fatal(err)
	}
	if len(listOfMachines) != 1 {
		logger.Fatal("Expected 1 machine after deletion")
	}

	fmt.Println("All tests passed successfully!")
}

// Helper function to create string pointers
func stringPtr(s string) *string {
	return &s
}

// Helper function to create int pointers
func intPtr(i int) *int {
	return &i
}

File: static/include/examples/fleet-api/fleet-management-api-machine.py
import asyncio
import os
import time

from viam.rpc.dial import DialOptions, Credentials
from viam.app.viam_client import ViamClient
from viam.proto.app import Authorization
from viam.proto.app import AuthorizedPermissions
from viam.proto.app.packages import PackageType
from viam.proto.app import Visibility, RegistryItemStatus
from viam.proto.app import Model, ModuleFileInfo
from viam.app.app_client import APIKeyAuthorization

# Configuration constants – replace with your actual values
API_KEY = ""  # API key, find or create in your organization settings
API_KEY_ID = ""  # API key ID, find or create in your organization settings
ORG_ID = ""  # Organization ID, find or create in your organization settings
EMAIL_ADDRESS = ""  # Email address of the user to get the user id for
LOCATION_ID = ""  # Location ID, find or create in your organization settings

# :remove-start:
ORG_ID = os.environ["TEST_ORG_ID"]
ORG_ID_2 = "b5e9f350-cbcf-4d2a-bbb1-a2e2fd6851e1"
API_KEY = os.environ["VIAM_API_KEY"]
API_KEY_ID = os.environ["VIAM_API_KEY_ID"]
LOCATION_ID = "pg5q3j3h95"
TEST_EMAIL = os.environ["TEST_EMAIL"]
MACHINE_ID = "5ec7266e-f762-4ea8-9c29-4dd592718b48"
PART_ID = "deb8782c-7b48-4d35-812d-2caa94b61f77"
# :remove-end:

async def connect() -> ViamClient:
    dial_options = DialOptions(
      credentials=Credentials(
        type="api-key",
        # Replace "<API-KEY>" (including brackets) with your machine's API key
        payload=API_KEY,
      ),
      # Replace "<API-KEY-ID>" (including brackets) with your machine's
      # API key ID
      auth_entity=API_KEY_ID
    )
    return await ViamClient.create_from_dial_options(dial_options)

async def main():

    # Make a ViamClient
    viam_client = await connect()
    # Instantiate an AppClient called "cloud"
    # to run fleet management API methods on
    cloud = viam_client.app_client

    robot = await cloud.get_robot(robot_id=MACHINE_ID)
    assert robot.id == MACHINE_ID

    api_keys = await cloud.get_robot_api_keys(robot_id=MACHINE_ID)
    assert len(api_keys) >= 1
    assert api_keys[0].authorizations[0].resource_id == MACHINE_ID

    list_of_parts = await cloud.get_robot_parts(
        robot_id=MACHINE_ID
    )
    assert len(list_of_parts) == 1
    assert list_of_parts[0].id == PART_ID

    my_robot_part = await cloud.get_robot_part(
      robot_part_id=PART_ID)
    assert my_robot_part.id == PART_ID

    part_logs = await cloud.get_robot_part_logs(
        robot_part_id=PART_ID, num_log_entries=5
    )
    assert len(part_logs) == 5

    logs_stream = await cloud.tail_robot_part_logs(
        robot_part_id=PART_ID
    )
    assert logs_stream is not None

    part_history = await cloud.get_robot_part_history(
        robot_part_id=PART_ID
    )
    assert len(part_history) >= 1

    new_part_id = await cloud.new_robot_part(
        robot_id=MACHINE_ID, part_name="new-part"
    )
    assert new_part_id is not None
    list_of_parts = await cloud.get_robot_parts(
        robot_id=MACHINE_ID
    )
    assert len(list_of_parts) == 2

    await cloud.mark_part_as_main(
        robot_part_id=new_part_id
    )
    new_machine_part = await cloud.get_robot_part(
        robot_part_id=new_part_id
    )
    assert new_machine_part.main_part == True
    await cloud.mark_part_as_main(
        robot_part_id=PART_ID
    )
    old_machine_part = await cloud.get_robot_part(
        robot_part_id=PART_ID
    )
    assert old_machine_part.main_part == True

    await cloud.mark_part_for_restart(
        robot_part_id=new_part_id
    )

    part_with_new_secret = await cloud.create_robot_part_secret(
        robot_part_id=new_part_id
    )
    assert part_with_new_secret is not None
    assert part_with_new_secret.id == new_part_id

    await cloud.delete_robot_part_secret(
        robot_part_id=new_part_id,
        secret_id=part_with_new_secret.secrets[0].id)

    await cloud.delete_robot_part(
        robot_part_id=new_part_id
    )
    list_of_parts = await cloud.get_robot_parts(
        robot_id=MACHINE_ID
    )
    assert len(list_of_parts) == 1


    list_of_machines = await cloud.list_robots(location_id=LOCATION_ID)
    len_machines = len(list_of_machines)
    assert len_machines >= 1
    assert list_of_machines[0].id == MACHINE_ID

    new_machine_id = await cloud.new_robot(name="test-robot", location_id=LOCATION_ID)
    assert new_machine_id is not None
    list_of_machines = await cloud.list_robots(location_id=LOCATION_ID)
    assert len(list_of_machines) == len_machines + 1

    updated_machine = await cloud.update_robot(
      robot_id=new_machine_id,
      name="test-robot-new-name",
      location_id=LOCATION_ID
    )
    assert updated_machine.name == "test-robot-new-name"

    await cloud.delete_robot(robot_id=new_machine_id)
    list_of_machines = await cloud.list_robots(location_id=LOCATION_ID)
    assert len(list_of_machines) == len_machines

    viam_client.close()

if __name__ == '__main__':
    asyncio.run(main())
